### ISSUE TITLE
Flash Attention in GPT-NeoX(LLM Training) Adaptation

### ISSUE THREAD

**Issue by fecet** [Dao-AILab/flash-attention#478]
Hello Flash Attention Team,

First and foremost, I want to express my deepest gratitude and admiration for your work on flash attention. The innovations and contributions you have made are genuinely impressive.

I've been working on adding tensor-parallel self attention flash attention2 support for `gpt-neox`. For this purpose, I've primarily been utilizing `flash_attn_func`, and after adapting the shapes and computation process to `neox`, it runs smoothly. However, I have run into an unexpected issue. The training results with flash attention are considerably different than those without it. This same discrepancy was observed when using the earlier versions of flash attention (1.0.x/0.x) with `gpt-neox`. As illustrated in the attached graph, the loss plummets to near 0 almost immediately when using flash attention. In contrast, the expected curve, shown in gray, utilizes `torch.baddbmm` and `fused_scaled_softmax` with `causal_attention_mask`.

![image](https://github.com/Dao-AILab/flash-attention/assets/41792945/3713feb2-7b17-41be-8277-9a55a758c30f)

During my training and testing, I employed DeepSpeed ZeRO2 optimizer and PyTorch gradient checkpointing. Is there any known incompatibility between these components and flash attention?

For clarity, here's the code of my adaptation:
```python
""" We use the following notation throughout this file:
     h: hidden size
     n: number of attention heads
     p: number of model parallel partitions
     np: n/p
     hp: h/p
     hn: h/n
     b: batch size
     s: sequence length
     l: number of layers
    Transformer takes input of size [s, b, h] and returns a
    tensor of the same size. We use the following arguments:
        hyperparameters: transformer hyperparameters
        attention_mask_func: a function that takes `unmasked-attention-scores`
            with size [b, np, s, s] and an `attention-mask` and will apply
            the masking. The function should return a masked score of the
            same size [b, np, s, s].
               masked-attention-scores = attention_mask_func(
                                     unmasked-attention-scores, attention-mask)
"""

...

if flash_attn_version >= (2, 0, 0):
    from flash_attn import (
        flash_attn_qkvpacked_func,
        flash_attn_func,
    )

    self.flash_qkv_fn = flash_attn_qkvpacked_func
    self.flash_attn_fn = flash_attn_func
    self.use_flash_attn2 = True

....

# [b, np, sq, sk]
output_size = (
    query_layer.size(1),
    query_layer.size(2),
    query_layer.size(0),
    key_layer.size(0),
)

if self.training:
    # [sq, b, np, hn] -> [b, sq, np, hn]
    query_layer = query_layer.transpose(0, 1).reshape(
        output_size[0], output_size[2], output_size[1], -1
    )
    key_layer = query_layer.transpose(0, 1).reshape(
        output_size[0], output_size[2], output_size[1], -1
    )
    value_layer = query_layer.transpose(0, 1).reshape(
        output_size[0], output_size[2], output_size[1], -1
    )

    # Combined q/k/v into [b, s, 3, np, hn].
    # qkv = torch.cat([query_layer, key_layer, value_layer], dim=2)

    # output = self.flash_qkv_fn(
    #     qkv,
    #     self.dropout_p if self.training else 0.0,
    #     softmax_scale=None,
    #     causal=True,
    # )
    output = self.flash_attn_fn(
        query_layer,
        key_layer,
        value_layer,
        self.dropout_p if self.training else 0.0,
        causal=True,
    )
    # qkv: (batch_size, seqlen, nheads, headdim) each
    # out: (batch_size, seqlen, nheads, headdim).
    matmul_result = output.view(
        output_size[0], output_size[2], output.shape[2], output.shape[3]
    )
    # [b, sq, np, hn] -> [b, np, sq, hn]
    matmul_result = matmul_result.transpose(1, 2)
    return matmul_result
else:
    raise NotImplementedError(
        "Flash attention 2 only implemented for training"
    )

```

**Comment by tridao** [comment]
It's possible that input has shape (seqlen, batch, ...) and you're treating it as (batch, seqlen, ...)? At least for Megatron they use seqlen first, so I had to transpose to have batch first. Idk what format gpt-neox uses.

You can set breakpoint and step through the code (standard attention and flashattention) to make sure the two give similar results (up to numerical error).

**Comment by fecet** [comment]
Sorry for me and my stupid Copilot-LSP who gave me joker code below
```python
if self.training:
    # [sq, b, np, hn] -> [b, sq, np, hn]
    query_layer = query_layer.transpose(0, 1).reshape(
        output_size[0], output_size[2], output_size[1], -1
    )
    key_layer = query_layer.transpose(0, 1).reshape(
        output_size[0], output_size[2], output_size[1], -1
    )
    value_layer = query_layer.transpose(0, 1).reshape(
        output_size[0], output_size[2], output_size[1], -1
    )
```
after changing `query_layer` to `key_layer` and `value_layer`, all results match.
Sorry again for wasting your time. I will try to commit to NeoX to support flash attention as soon as I can.
